rust/hyperlane-base/src/metrics/core.rs

```diff
@@ -468,3 +472,94 @@ impl Debug for CoreMetrics {
         )
     }
 }
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+struct AppContextKey {
+    origin: HyperlaneDomain,
+    destination: HyperlaneDomain,
+    app_context: String,
+}
+
+/// Manages metrics for observing sets of validators.
+pub struct ValidatorObservabilityMetricManager {
+    observed_validator_latest_index: IntGaugeVec,
+
+    app_context_validators: RwLock<HashMap<AppContextKey, HashSet<H160>>>,
+}
+
+impl ValidatorObservabilityMetricManager {
+    fn new(observed_validator_latest_index: IntGaugeVec) -> Self {
+        Self {
+            observed_validator_latest_index,
+            app_context_validators: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Updates the metrics with the latest checkpoint index for each validator
+    /// in a given set.
+    pub async fn set_validator_latest_checkpoints(
+        &self,
+        origin: &HyperlaneDomain,
+        destination: &HyperlaneDomain,
+        app_context: String,
+        latest_checkpoints: &HashMap<H160, Option<u32>>,
+    ) {
+        let key = AppContextKey {
+            origin: origin.clone(),
+            destination: destination.clone(),
+            app_context: app_context.clone(),
+        };
+
+        let mut app_context_validators = self.app_context_validators.write().await;
+
+        // First, clear out all previous metrics for the app context.
+        // This is necessary because the set of validators may have changed.
+        if let Some(prev_validators) = app_context_validators.get(&key) {
+            for validator in prev_validators {
+                self.observed_validator_latest_index
+                    .remove_label_values(&[
```

Comment:
why not use `.reset()` instead of `.remove_label_values()`?

Reply:
I think `reset()` will remove all metrics. E.g. if we're tracking app A and app B, when we're updating the metrics for app A we'll inadvertently delete all the metrics for app B